Make displayed dice rolls span multiple lines and other combat UI improvements

asvitkine opened the pull request:
This change improves the display of dice rolls when there's a lot
of them, by wrapping the dice as opposed to showing them all in
one scrollable line. To accommodate the change, which can increase
the height of the containing components, some UI elements are
restructured to use an outer scrollpane - particularly the combat
and history views.

Also improves the battle window in a few ways:
  - If the screen is more than 1024x768, use 1024x768 size
  - The total hits line is now on top in a larger font
  - For each dice value, show number of hits and number of rolls
  - Show number of hits in red

## Functional Changes
<!-- Put an X next any that apply -->
[] New map or map update
[] New Feature
[X] Feature update or enhancement
[] Feature Removal
[] Code Cleanup or refactor
[] Configuration Change
[] Problem fix:  <!-- Link to bug issue or forum post here -->
[] Other:   <!-- Please specify -->

## Testing
Manual testing on both existing save games and with map edit to add different, large number of units to verify things are displayed correctly in different scenarios. Tested resize behavior in both the combat window and the history panel to make sure the dice rolls are correctly reflowed. Tested on different themes.

## Screens Shots

### Before (Combat)
<img width="912" alt="Screen Shot 2020-05-08 at 8 52 17 AM" src="https://user-images.githubusercontent.com/17648/81475350-8e53aa80-91d9-11ea-8a32-95515e80b19c.png">
Notice the vertical scrollbar and the fact there's no horizontal scrollbar even though the content is already too tall.

### After (Combat)
![Screen Shot 2020-05-20 at 9 40 50 AM](https://user-images.githubusercontent.com/17648/82453074-1198ba00-9a7e-11ea-8621-fecdd601fc6a.png)

Save game for testing the above (end the combat step to see the battle window):
[before_dice.tsvg.zip](https://github.com/triplea-game/triplea/files/4603589/before_dice.tsvg.zip)

Dark theme, 12-sided dice with lots of rolls to require scrolling:
![Screen Shot 2020-05-22 at 9 16 16 AM](https://user-images.githubusercontent.com/17648/82672872-f1e4cb80-9c0e-11ea-8413-c18255f42732.png)

### Before (History)
Different theme, showing the history panel:
<img width="378" alt="Screen Shot 2020-05-09 at 10 11 07 AM" src="https://user-images.githubusercontent.com/17648/81476068-c957dd00-91dd-11ea-9b90-8132adbcceaf.png">


### After (History)
Different theme, showing the history panel:
![Screen Shot 2020-05-20 at 7 28 49 PM](https://user-images.githubusercontent.com/17648/82507423-30c03780-9ad0-11ea-8675-d49a0b3e44e2.png)

Save game for testing the history view (open saved game and switch to history view and view the combat where Austria attacks Greece): 
[dom_arabia_10.tsvg.zip](https://github.com/triplea-game/triplea/files/4603622/dom_arabia_10.tsvg.zip)


<!--
  Uncomment the below and add any additional details that would be helpful for reviewers.
-->
<!--
## Additional Review Notes
-->

tvleavitt commented:
Design feedback: rolls @ 1, especially misses, are the least interesting
info. Typically, you just think of these as lucky hits (even though there's
really no distinction mathematically) that you can't "count on", and you're
focused on how your higher value units performed.

Could we reverse the display order and show "Total hits" first, then higher
value dice rolls in declining order? As with the previous set up, I'm going
to guess most people won't scroll to see the results.

The other useful / interesting bit of info is how many hits were generated
by each dice roll type... right now, you have to calculate this in your
head. I don't know if it is even tracked this way, but it would be great to
see a sub-total of this displayed as well. Maybe in parenthesis like
"Rolled at 4 (3 hits):"?

Thomas Leavitt

On Sat, May 9, 2020, 6:45 AM asvitkine <notifications@github.com> wrote:

> This change improves the display of dice rolls when there's a lot
> of them, by wrapping the dice as opposed to showing them all in
> one scrollable line. To accomodate the change, which can increase
> the height of the containing components, some UI elements are
> restructured to use an outer scrollpane - particularly the combat
> and history views.
> Functional Changes
>
> [] New map or map update
> [] New Feature
> [X] Feature update or enhancement
> [] Feature Removal
> [] Code Cleanup or refactor
> [] Configuration Change
> [] Problem fix:
> [] Other:
> Testing
>
> Manual testing on both existing save games and with map edit to add
> different, large number of units to verify things are displayed correctly
> in different scenarios. Tested resize behavior in both the combat window
> and the history panel to make sure the dice rolls are correctly reflowed.
> Screens Shots Before
>
> [image: Screen Shot 2020-05-08 at 8 52 17 AM]
> <https://user-images.githubusercontent.com/17648/81475350-8e53aa80-91d9-11ea-8a32-95515e80b19c.png>
> After
>
> [image: Screen Shot 2020-05-07 at 12 26 25 AM]
> <https://user-images.githubusercontent.com/17648/81475362-ae836980-91d9-11ea-95ef-787fe5a614dd.png>
> [image: Screen Shot 2020-05-07 at 12 26 28 AM]
> <https://user-images.githubusercontent.com/17648/81475367-b6430e00-91d9-11ea-8421-5c0508f02cd0.png>
> [image: Screen Shot 2020-05-07 at 12 26 35 AM]
> <https://user-images.githubusercontent.com/17648/81475370-be02b280-91d9-11ea-8130-93da0fe73631.png>
>
> -->
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/triplea-game/triplea/pull/6449
> Commit Summary
>
>    - Make displayed dice rolls span multiple lines via a WrapLayout.
>
> File Changes
>
>    - *M*
>    game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java
>    <https://github.com/triplea-game/triplea/pull/6449/files#diff-db00a0dbf155c0e7790aecf71bb138ef>
>    (83)
>    - *M* game-core/src/main/java/games/strategy/triplea/ui/DicePanel.java
>    <https://github.com/triplea-game/triplea/pull/6449/files#diff-44d3427d98ffc5baa496ccf1423cfdb7>
>    (58)
>    - *M*
>    game-core/src/main/java/games/strategy/triplea/ui/history/HistoryDetailsPanel.java
>    <https://github.com/triplea-game/triplea/pull/6449/files#diff-a26a54aad1a82b5ce2ea6fe63856d1ac>
>    (53)
>    - *A* swing-lib/src/main/java/org/triplea/swing/ScrollableJPanel.java
>    <https://github.com/triplea-game/triplea/pull/6449/files#diff-81f62c3babc4ae968b0591f4da24c708>
>    (56)
>    - *M* swing-lib/src/main/java/org/triplea/swing/WrapLayout.java
>    <https://github.com/triplea-game/triplea/pull/6449/files#diff-81fba2c6332e047ec8bd1793749b9718>
>    (31)
>
> Patch Links:
>
>    - https://github.com/triplea-game/triplea/pull/6449.patch
>    - https://github.com/triplea-game/triplea/pull/6449.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB5HAMTTPW7XQRAGJGKVCTRQVM7TANCNFSM4M4ZYJKA>
> .
>

asvitkine commented:
Those are good points. I'm not sure it's best to do them in this PR though
as opposed to a separate focused change.

Although the simple one of showing the number of hits in the label would be
easy to add here without major changes.

On Sat, May 9, 2020, 10:25 AM Thomas Leavitt <notifications@github.com>
wrote:

> Design feedback: rolls @ 1, especially misses, are the least interesting
> info. Typically, you just think of these as lucky hits (even though there's
> really no distinction mathematically) that you can't "count on", and you're
> focused on how your higher value units performed.
>
> Could we reverse the display order and show "Total hits" first, then higher
> value dice rolls in declining order? As with the previous set up, I'm going
> to guess most people won't scroll to see the results.
>
> The other useful / interesting bit of info is how many hits were generated
> by each dice roll type... right now, you have to calculate this in your
> head. I don't know if it is even tracked this way, but it would be great to
> see a sub-total of this displayed as well. Maybe in parenthesis like
> "Rolled at 4 (3 hits):"?
>
> Thomas Leavitt
>
> On Sat, May 9, 2020, 6:45 AM asvitkine <notifications@github.com> wrote:
>
> > This change improves the display of dice rolls when there's a lot
> > of them, by wrapping the dice as opposed to showing them all in
> > one scrollable line. To accomodate the change, which can increase
> > the height of the containing components, some UI elements are
> > restructured to use an outer scrollpane - particularly the combat
> > and history views.
> > Functional Changes
> >
> > [] New map or map update
> > [] New Feature
> > [X] Feature update or enhancement
> > [] Feature Removal
> > [] Code Cleanup or refactor
> > [] Configuration Change
> > [] Problem fix:
> > [] Other:
> > Testing
> >
> > Manual testing on both existing save games and with map edit to add
> > different, large number of units to verify things are displayed correctly
> > in different scenarios. Tested resize behavior in both the combat window
> > and the history panel to make sure the dice rolls are correctly reflowed.
> > Screens Shots Before
> >
> > [image: Screen Shot 2020-05-08 at 8 52 17 AM]
> > <
> https://user-images.githubusercontent.com/17648/81475350-8e53aa80-91d9-11ea-8a32-95515e80b19c.png
> >
> > After
> >
> > [image: Screen Shot 2020-05-07 at 12 26 25 AM]
> > <
> https://user-images.githubusercontent.com/17648/81475362-ae836980-91d9-11ea-95ef-787fe5a614dd.png
> >
> > [image: Screen Shot 2020-05-07 at 12 26 28 AM]
> > <
> https://user-images.githubusercontent.com/17648/81475367-b6430e00-91d9-11ea-8421-5c0508f02cd0.png
> >
> > [image: Screen Shot 2020-05-07 at 12 26 35 AM]
> > <
> https://user-images.githubusercontent.com/17648/81475370-be02b280-91d9-11ea-8130-93da0fe73631.png
> >
> >
> > -->
> > ------------------------------
> > You can view, comment on, or merge this pull request online at:
> >
> > https://github.com/triplea-game/triplea/pull/6449
> > Commit Summary
> >
> > - Make displayed dice rolls span multiple lines via a WrapLayout.
> >
> > File Changes
> >
> > - *M*
> > game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java
> > <
> https://github.com/triplea-game/triplea/pull/6449/files#diff-db00a0dbf155c0e7790aecf71bb138ef
> >
> > (83)
> > - *M* game-core/src/main/java/games/strategy/triplea/ui/DicePanel.java
> > <
> https://github.com/triplea-game/triplea/pull/6449/files#diff-44d3427d98ffc5baa496ccf1423cfdb7
> >
> > (58)
> > - *M*
> >
> game-core/src/main/java/games/strategy/triplea/ui/history/HistoryDetailsPanel.java
> > <
> https://github.com/triplea-game/triplea/pull/6449/files#diff-a26a54aad1a82b5ce2ea6fe63856d1ac
> >
> > (53)
> > - *A* swing-lib/src/main/java/org/triplea/swing/ScrollableJPanel.java
> > <
> https://github.com/triplea-game/triplea/pull/6449/files#diff-81f62c3babc4ae968b0591f4da24c708
> >
> > (56)
> > - *M* swing-lib/src/main/java/org/triplea/swing/WrapLayout.java
> > <
> https://github.com/triplea-game/triplea/pull/6449/files#diff-81fba2c6332e047ec8bd1793749b9718
> >
> > (31)
> >
> > Patch Links:
> >
> > - https://github.com/triplea-game/triplea/pull/6449.patch
> > - https://github.com/triplea-game/triplea/pull/6449.diff
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/triplea-game/triplea/pull/6449>, or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABB5HAMTTPW7XQRAGJGKVCTRQVM7TANCNFSM4M4ZYJKA
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-626184010>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEJ4C4CRDTHINVW57DBYLRQVRUZANCNFSM4M4ZYJKA>
> .
>

tvleavitt commented:
I think best practices would be to make additional changes outside the
scope of the PR in a separate PR, yes. Granularity as am operating
principle.

Thomas Leavitt

On Sat, May 9, 2020, 9:28 AM asvitkine <notifications@github.com> wrote:

> Those are good points. I'm not sure it's best to do them in this PR though
> as opposed to a separate focused change.
>
> Although the simple one of showing the number of hits in the label would be
> easy to add here without major changes.
>
> On Sat, May 9, 2020, 10:25 AM Thomas Leavitt <notifications@github.com>
> wrote:
>
> > Design feedback: rolls @ 1, especially misses, are the least interesting
> > info. Typically, you just think of these as lucky hits (even though
> there's
> > really no distinction mathematically) that you can't "count on", and
> you're
> > focused on how your higher value units performed.
> >
> > Could we reverse the display order and show "Total hits" first, then
> higher
> > value dice rolls in declining order? As with the previous set up, I'm
> going
> > to guess most people won't scroll to see the results.
> >
> > The other useful / interesting bit of info is how many hits were
> generated
> > by each dice roll type... right now, you have to calculate this in your
> > head. I don't know if it is even tracked this way, but it would be great
> to
> > see a sub-total of this displayed as well. Maybe in parenthesis like
> > "Rolled at 4 (3 hits):"?
> >
> > Thomas Leavitt
> >
> > On Sat, May 9, 2020, 6:45 AM asvitkine <notifications@github.com> wrote:
> >
> > > This change improves the display of dice rolls when there's a lot
> > > of them, by wrapping the dice as opposed to showing them all in
> > > one scrollable line. To accomodate the change, which can increase
> > > the height of the containing components, some UI elements are
> > > restructured to use an outer scrollpane - particularly the combat
> > > and history views.
> > > Functional Changes
> > >
> > > [] New map or map update
> > > [] New Feature
> > > [X] Feature update or enhancement
> > > [] Feature Removal
> > > [] Code Cleanup or refactor
> > > [] Configuration Change
> > > [] Problem fix:
> > > [] Other:
> > > Testing
> > >
> > > Manual testing on both existing save games and with map edit to add
> > > different, large number of units to verify things are displayed
> correctly
> > > in different scenarios. Tested resize behavior in both the combat
> window
> > > and the history panel to make sure the dice rolls are correctly
> reflowed.
> > > Screens Shots Before
> > >
> > > [image: Screen Shot 2020-05-08 at 8 52 17 AM]
> > > <
> >
> https://user-images.githubusercontent.com/17648/81475350-8e53aa80-91d9-11ea-8a32-95515e80b19c.png
> > >
> > > After
> > >
> > > [image: Screen Shot 2020-05-07 at 12 26 25 AM]
> > > <
> >
> https://user-images.githubusercontent.com/17648/81475362-ae836980-91d9-11ea-95ef-787fe5a614dd.png
> > >
> > > [image: Screen Shot 2020-05-07 at 12 26 28 AM]
> > > <
> >
> https://user-images.githubusercontent.com/17648/81475367-b6430e00-91d9-11ea-8421-5c0508f02cd0.png
> > >
> > > [image: Screen Shot 2020-05-07 at 12 26 35 AM]
> > > <
> >
> https://user-images.githubusercontent.com/17648/81475370-be02b280-91d9-11ea-8130-93da0fe73631.png
> > >
> > >
> > > -->
> > > ------------------------------
> > > You can view, comment on, or merge this pull request online at:
> > >
> > > https://github.com/triplea-game/triplea/pull/6449
> > > Commit Summary
> > >
> > > - Make displayed dice rolls span multiple lines via a WrapLayout.
> > >
> > > File Changes
> > >
> > > - *M*
> > > game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java
> > > <
> >
> https://github.com/triplea-game/triplea/pull/6449/files#diff-db00a0dbf155c0e7790aecf71bb138ef
> > >
> > > (83)
> > > - *M* game-core/src/main/java/games/strategy/triplea/ui/DicePanel.java
> > > <
> >
> https://github.com/triplea-game/triplea/pull/6449/files#diff-44d3427d98ffc5baa496ccf1423cfdb7
> > >
> > > (58)
> > > - *M*
> > >
> >
> game-core/src/main/java/games/strategy/triplea/ui/history/HistoryDetailsPanel.java
> > > <
> >
> https://github.com/triplea-game/triplea/pull/6449/files#diff-a26a54aad1a82b5ce2ea6fe63856d1ac
> > >
> > > (53)
> > > - *A* swing-lib/src/main/java/org/triplea/swing/ScrollableJPanel.java
> > > <
> >
> https://github.com/triplea-game/triplea/pull/6449/files#diff-81f62c3babc4ae968b0591f4da24c708
> > >
> > > (56)
> > > - *M* swing-lib/src/main/java/org/triplea/swing/WrapLayout.java
> > > <
> >
> https://github.com/triplea-game/triplea/pull/6449/files#diff-81fba2c6332e047ec8bd1793749b9718
> > >
> > > (31)
> > >
> > > Patch Links:
> > >
> > > - https://github.com/triplea-game/triplea/pull/6449.patch
> > > - https://github.com/triplea-game/triplea/pull/6449.diff
> > >
> > > —
> > > You are receiving this because you are subscribed to this thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/triplea-game/triplea/pull/6449>, or unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/ABB5HAMTTPW7XQRAGJGKVCTRQVM7TANCNFSM4M4ZYJKA
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/triplea-game/triplea/pull/6449#issuecomment-626184010>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAAEJ4C4CRDTHINVW57DBYLRQVRUZANCNFSM4M4ZYJKA
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-626201389>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB5HALC7JYHBVSLGPAHEC3RQWADLANCNFSM4M4ZYJKA>
> .
>

DanVanAtta commented:
Having a tally per dice roll of hits is a very good suggestion & update.

@asvitkine have you considered and what are your thoughts on cases with *very* large armies, potentially with 12 sided dice? For example WaW can see 150+ units per side in battles. I'm not aware of any 12 sided dice games that typically have large armies, but as an engine we need to be flexible for many possible varieties of maps. 

Would it make sense to cap the number rows to an upper limit, perhaps something like 3 or 4? If there are many dice, would rendering the dice roll images smaller be of help?

Have you had thoughts or considered making individual dice roll areas scrollable? In this way if there are 100 rolls against a 1, you can see maybe something like a 5x5 grid that can vertically scroll.

Cernelius commented:
In my mind, having a single line for each dice power makes the most sense, but I agree missing the vertical scroll bar was a bad US. You should not be obliged to resize to see everything (since also you may have so much stuff that you can just have it on the screen, in theory). Talking about this, you should also not be obliged to resize, to see the fighter in the killed line, in my opinion.

Cernelius commented:
Of course, the dice can be customized. I believe each dice can have a custom image of whatever size, even though, if I recall correctly, they are cut if higher than 48 pixels or close. Maybe you should take some screenshot of something like customized 64x64 pixels dice images.

asvitkine commented:
In my screenshot above, there's over 100 rolls at 1. So it handles it just fine.

Note that most smaller battles would be either completely the same as before (if fitting on one line) or just a couple lines longer.

Having said that, I do like the idea of scaling the dice down if there are many. Do we have smaller images or should the engine just rescale them itself?

Cernelius commented:
I neither had nor proposed any ideas about scaling down dice images. Thinking about this for the first time, I don't like the idea (unless maybe if we are dealing with vector images, which we are not).

I'm not aware multiple images for a same dice result are supported.

To clarify, I prefer the single line distribution, since this is akin to have a table with the power rolls in the rows and the dice rolled in the columns, but your disposition is fine too, since the dice rolled for a same power are sets, not sequences (maybe the dice images should be centred, then).

What I certainly like is the addition of the vertical scroll bar. I wonder if it would have been possible adding a vertical scroll bar while keeping the horizontal scroll bars, for the single lines of dice results (as I would prefer).

Cernelius commented:
> In my screenshot above, there's over 100 rolls at 1. So it handles it just fine.

I could argue that this is merely due to the fact that the default dice offered by TripleA are too small, thereby you have way too much space for them. Hence, I may be making a map with much bigger dice. In the moment dice are (thankfully) customizable, I'm not sure the default really matters much.

DanVanAtta commented:
We perhaps should be scaling discussion to the side (but I'd like to keep it going in another thread as it is interesting. Having many many dice and scaling them so that more can be seen is IMO impressive. If we standardize that dice should be 48 or 64 or equivalent, then scaling them down by 50% should be lossless).

What are thoughts on having scroll bar areas for individual dice roll areas? IE: All the 1's have a scrollbar area, all the 2's, etc...  In this way you would not have to scroll to see how many units hit at '4' for example and you should always be able to see the total hits without scrolling (which I think is significant; players *do* want to see total hits, that is perhaps most important and the rest is sugar). Having to scroll to see casualties is perhaps also not ideal. I think we maybe could move that stuff above the dice rolls to avoid scrolling issues, IMHO scrolling each dice area gives perhaps the best of both worlds. So essence we would have largely the same behavior as before, but instead of 1 line of dice rolls there would be 3~5 and it would vertically scroll instead of horizontally. If we are fitting 15 or 20 dice rolls across, then it's only an army group of more than 60 before we start to see any vertical scrolling, which should guarantee that you always get to see the most needed information and the times to scroll would be quite limited.

asvitkine commented:
How do you decide what height to default each scroll area to?

Honestly I'm not sure if the problem wouldn't be better solved by making
the window larger by default (maybe checking the resolution to only do that
if monitor is big enough).

On Sun, May 10, 2020, 3:21 PM Dan Van Atta <notifications@github.com> wrote:

> We perhaps should be scaling discussion to the side (but I'd like to keep
> it going in another thread as it is interesting. Having many many dice and
> scaling them so that more can be seen is IMO impressive. If we standardize
> that dice should be 48 or 64 or equivalent, then scaling them down by 50%
> should be lossless).
>
> What are thoughts on having scroll bar areas for individual dice roll
> areas? IE: All the 1's have a scrollbar area, all the 2's, etc... In this
> way you would not have to scroll to see how many units hit at '4' for
> example and you should always be able to see the total hits without
> scrolling (which I think is significant; players *do* want to see total
> hits, that is perhaps most important and the rest is sugar). Having to
> scroll to see casualties is perhaps also not ideal. I think we maybe could
> move that stuff above the dice rolls to avoid scrolling issues, IMHO
> scrolling each dice area gives perhaps the best of both worlds. So essence
> we would have largely the same behavior as before, but instead of 1 line of
> dice rolls there would be 3~5 and it would vertically scroll instead of
> horizontally. If we are fitting 15 or 20 dice rolls across, then it's only
> an army group of more than 60 before we start to see any vertical
> scrolling, which should guarantee that you always get to see the most
> needed information and the times to scroll would be quite limited.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-626375971>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEJ4E5Q7MNDOV6HEA2VK3RQ35CVANCNFSM4M4ZYJKA>
> .
>

DanVanAtta commented:
The height calculation perhaps would be:

Rows*(dice image height) + (any padding)

A bigger window by about 20% perhaps is not a bad additional idea.

On Sun, May 10, 2020, 1:13 PM asvitkine <notifications@github.com> wrote:

> How do you decide what height to default each scroll area to?
>
> Honestly I'm not sure if the problem wouldn't be better solved by making
> the window larger by default (maybe checking the resolution to only do that
> if monitor is big enough).
>
> On Sun, May 10, 2020, 3:21 PM Dan Van Atta <notifications@github.com>
> wrote:
>
> > We perhaps should be scaling discussion to the side (but I'd like to keep
> > it going in another thread as it is interesting. Having many many dice
> and
> > scaling them so that more can be seen is IMO impressive. If we
> standardize
> > that dice should be 48 or 64 or equivalent, then scaling them down by 50%
> > should be lossless).
> >
> > What are thoughts on having scroll bar areas for individual dice roll
> > areas? IE: All the 1's have a scrollbar area, all the 2's, etc... In this
> > way you would not have to scroll to see how many units hit at '4' for
> > example and you should always be able to see the total hits without
> > scrolling (which I think is significant; players *do* want to see total
> > hits, that is perhaps most important and the rest is sugar). Having to
> > scroll to see casualties is perhaps also not ideal. I think we maybe
> could
> > move that stuff above the dice rolls to avoid scrolling issues, IMHO
> > scrolling each dice area gives perhaps the best of both worlds. So
> essence
> > we would have largely the same behavior as before, but instead of 1 line
> of
> > dice rolls there would be 3~5 and it would vertically scroll instead of
> > horizontally. If we are fitting 15 or 20 dice rolls across, then it's
> only
> > an army group of more than 60 before we start to see any vertical
> > scrolling, which should guarantee that you always get to see the most
> > needed information and the times to scroll would be quite limited.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/triplea-game/triplea/pull/6449#issuecomment-626375971>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAAEJ4E5Q7MNDOV6HEA2VK3RQ35CVANCNFSM4M4ZYJKA
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-626382908>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC6SZONNUBEIQUH6PDPPCPTRQ4DHBANCNFSM4M4ZYJKA>
> .
>

asvitkine commented:
> The height calculation perhaps would be: Rows*(dice image height) + (any padding) A bigger window by about 20% perhaps is not a bad additional idea.
> […](#)
> On Sun, May 10, 2020, 1:13 PM asvitkine ***@***.***> wrote: How do you decide what height to default each scroll area to? Honestly I'm not sure if the problem wouldn't be better solved by making the window larger by default (maybe checking the resolution to only do that if monitor is big enough). On Sun, May 10, 2020, 3:21 PM Dan Van Atta ***@***.***> wrote: > We perhaps should be scaling discussion to the side (but I'd like to keep > it going in another thread as it is interesting. Having many many dice and > scaling them so that more can be seen is IMO impressive. If we standardize > that dice should be 48 or 64 or equivalent, then scaling them down by 50% > should be lossless). > > What are thoughts on having scroll bar areas for individual dice roll > areas? IE: All the 1's have a scrollbar area, all the 2's, etc... In this > way you would not have to scroll to see how many units hit at '4' for > example and you should always be able to see the total hits without > scrolling (which I think is significant; players *do* want to see total > hits, that is perhaps most important and the rest is sugar). Having to > scroll to see casualties is perhaps also not ideal. I think we maybe could > move that stuff above the dice rolls to avoid scrolling issues, IMHO > scrolling each dice area gives perhaps the best of both worlds. So essence > we would have largely the same behavior as before, but instead of 1 line of > dice rolls there would be 3~5 and it would vertically scroll instead of > horizontally. If we are fitting 15 or 20 dice rolls across, then it's only > an army group of more than 60 before we start to see any vertical > scrolling, which should guarantee that you always get to see the most > needed information and the times to scroll would be quite limited. > > — > You are receiving this because you were mentioned. > Reply to this email directly, view it on GitHub > < [#6449 (comment)](https://github.com/triplea-game/triplea/pull/6449#issuecomment-626375971)>, > or unsubscribe > < https://github.com/notifications/unsubscribe-auth/AAAEJ4E5Q7MNDOV6HEA2VK3RQ35CVANCNFSM4M4ZYJKA > > . > — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#6449 (comment)](https://github.com/triplea-game/triplea/pull/6449#issuecomment-626382908)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AC6SZONNUBEIQUH6PDPPCPTRQ4DHBANCNFSM4M4ZYJKA> .

asvitkine commented:
> The height calculation perhaps would be: Rows*(dice image height) + (any padding)

One of my screenshots already showed the window being too short before my change with just 4 rows of dice shown + casualties. How do we handle that? Is the suggestion to make tall enough to always shows at 6 dice + casualties to start with? Or do we have two layers of scroll bars (yuck)?

DanVanAtta commented:
Scroll bar in scroll bar is bad, let's avoid. Extra window height could
help. It's part of the issue the window being too short, reducing rows in
theory helps. We could go dynamic maybe and show a max number of rows and
compute an averag re,  but seemingly that is getting complex.

On Sun, May 10, 2020, 2:38 PM asvitkine <notifications@github.com> wrote:

> The height calculation perhaps would be: Rows*(dice image height) + (any
> padding)
>
> One of my screenshots already showed the window being too short before my
> change with just 4 rows of dice shown + casualties. How do we handle that?
> Is the suggestion to make tall enough to always shows at 6 dice +
> casualties to start with? Or do we have two layers of scroll bars (yuck)?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-626393647>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC6SZON243X5VU6VWMV6TQTRQ4NGFANCNFSM4M4ZYJKA>
> .
>

DanVanAtta commented:
Another idea for you @asvitkine , perhaps we can make the casualty and total hits be a fixed panel and the rest scrollable.

Cernelius commented:
>Or do we have two layers of scroll bars (yuck)?

Yes, that was my suggestion. Basically a main vertical general scroll bar plus a series of horizontal scrollbars, one for each dice and one for the "killed" units display. Practically, just keeping the old layout, but adding the vertical scrollbar and an horizontal scroll bar for the "killed" units display.

Anyways, your initial proposal is good too, except only for the fact that the "killed" units should not be cut without scrollbar, but this was an issue all the same previously. One way or the other works well enough.

Really not a big deal anyways. I would be curious to see how it looks for 48x48 px and 64x64 px dice, though.

asvitkine commented:
Just as an update, I'm still planning to incorporate the feedback and see how it works. Just have been busy so might take a bit for me to polish things up.

asvitkine commented:
I applied some of the ideas in this thread. In particular:
 - Make the casualties panel not be part of the scroll view, but at a fixed location.
 - Show total hits at the top.
 - Show number of hits for each dice roll (and number of dice rolled).
 - Make the window 1024x768 instead of 800x600 if the screen size is above 1024x768.

I think it's a good number of improvements. Take a look at how it looks with my example of 100 dice at 1, which now doesn't even require scrolling with the 1024x768 size.

![Screen Shot 2020-05-16 at 9 55 32 AM](https://user-images.githubusercontent.com/17648/82121544-f1a28700-975b-11ea-9b2b-436846550c49.png)

Here's how it looks with a normal, small attack instead:
![Screen Shot 2020-05-16 at 9 55 38 AM](https://user-images.githubusercontent.com/17648/82121555-fff0a300-975b-11ea-985d-bad038250369.png)

Cernelius commented:
> Make the window 1024x768 instead of 800x600 if the screen size is above 1024x768.

I disagree setting to absolute values. I believe all values should be relative (in this case, a percentage of the screen size).

tvleavitt commented:
I agree, if the code change isn't too much of a hassle. I play on both a
1366x768 MBA, and a 2560x1080 discrete monitor / desktop combo. No reason
to arbitrarily limit the size in the latter context.

Thomas Leavitt

On Sat, May 16, 2020, 8:54 AM Cernelius <notifications@github.com> wrote:

> Make the window 1024x768 instead of 800x600 if the screen size is above
> 1024x768.
>
> I disagree setting to absolute values. I believe all values should be
> relative (in this case, a percentage of the screen size).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-629666818>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB5HAPTLMTTCY4LVP3EEC3RR2ZMHANCNFSM4M4ZYJKA>
> .
>

asvitkine commented:
The window is resizable though. It's the minimum size that we set.

What size would you prefer it defaulted to as the minimum on your large
monitor?

On Sat, May 16, 2020, 4:18 PM Thomas Leavitt <notifications@github.com>
wrote:

> I agree, if the code change isn't too much of a hassle. I play on both a
> 1366x768 MBA, and a 2560x1080 discrete monitor / desktop combo. No reason
> to arbitrarily limit the size in the latter context.
>
> Thomas Leavitt
>
> On Sat, May 16, 2020, 8:54 AM Cernelius <notifications@github.com> wrote:
>
> > Make the window 1024x768 instead of 800x600 if the screen size is above
> > 1024x768.
> >
> > I disagree setting to absolute values. I believe all values should be
> > relative (in this case, a percentage of the screen size).
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/triplea-game/triplea/pull/6449#issuecomment-629666818>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABB5HAPTLMTTCY4LVP3EEC3RR2ZMHANCNFSM4M4ZYJKA
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-629700685>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEJ4B5SRWMXB4NZFUA6W3RR3YIHANCNFSM4M4ZYJKA>
> .
>

tvleavitt commented:
I think that's an improvement with respect to minimum size; I'd survey the
folks on the forums for more a more general take. I think this depends, in
large part, on whether you run TripleA in a window, or fullscreen... in
full screen, 1024 represents less than 50% of the width of the screen,
plenty of room for a larger window. I think in fullscreen, you could
default to approximately 70% width and 70% height safely. 1792x768 would do
it for my screen. Others have screens that run at 1440 pixels in height,
which makes the battle window tiny (less than 40% with and barely 50%
height). In a window, it's different. What's the default window size?

Is there an issue with dynamically setting the minimum size, coding wise?
Is it a hassle to dynamically query the window size and set it to a
percentage of that?

I know it's a huge ask, but maybe even making that a settings option? I'm
not going to be resizing the window more than once in a blue moon, just
living with the size, and I think that'll be most people's habits...

Thomas

On Sat, May 16, 2020 at 1:25 PM asvitkine <notifications@github.com> wrote:

> The window is resizable though. It's the minimum size that we set.
>
> What size would you prefer it defaulted to as the minimum on your large
> monitor?
>
> On Sat, May 16, 2020, 4:18 PM Thomas Leavitt <notifications@github.com>
> wrote:
>
> > I agree, if the code change isn't too much of a hassle. I play on both a
> > 1366x768 MBA, and a 2560x1080 discrete monitor / desktop combo. No reason
> > to arbitrarily limit the size in the latter context.
> >
> > Thomas Leavitt
> >
> > On Sat, May 16, 2020, 8:54 AM Cernelius <notifications@github.com>
> wrote:
> >
> > > Make the window 1024x768 instead of 800x600 if the screen size is above
> > > 1024x768.
> > >
> > > I disagree setting to absolute values. I believe all values should be
> > > relative (in this case, a percentage of the screen size).
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <
> > https://github.com/triplea-game/triplea/pull/6449#issuecomment-629666818
> >,
> > > or unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/ABB5HAPTLMTTCY4LVP3EEC3RR2ZMHANCNFSM4M4ZYJKA
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/triplea-game/triplea/pull/6449#issuecomment-629700685>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAAEJ4B5SRWMXB4NZFUA6W3RR3YIHANCNFSM4M4ZYJKA
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-629701451>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB5HAPMPQ2FJMIDSW43V2DRR3ZEFANCNFSM4M4ZYJKA>
> .
>


-- 
Thomas Leavitt
Internet enabled since 1990

asvitkine commented:
Rather than making an option, how about remembering the last window size? So when it's opened again, it will be the same size as before. I can look at doing that.

Cernelius commented:
> Rather than making an option, how about remembering the last window size? So when it's opened again, it will be the same size as before. I can look at doing that.

I don't like the idea, unless there is any easy to figure way to reformat it to default size.

tvleavitt commented:
Hmm... on first blush, leaving it at the last size sounds reasonable, if
you're going to be resizing it from the default because it is just too
small. On the other hand, resizing it from there is likely to only be done
in exceptional circumstances (such as loads of dice), and it make sense to
automatically revert back from that default size... but then, it isn't that
much hassle to simply shrink a window again, and certainly, adding a UI
element to either set the default size or revert it back complicates the
interface and coding for something likely to be relatively infrequently
used for which there is an adequate generic alterantive. @Cernelius -
that'd be the way to "reformat", just use the OS built in window sizing
widget built into every window. I think? So, pretty much zero hassle in
that respect.

On balance, I think that's the right approach. The question is, how do we
communicate to the end user that permanently resizing the battle window is
an option? Embed it in the tutorial? Have it as a tooltip somehow? Or just
letting people discover it is an option by doing it?

Thomas

On Mon, May 18, 2020 at 11:48 AM Cernelius <notifications@github.com> wrote:

> Rather than making an option, how about remembering the last window size?
> So when it's opened again, it will be the same size as before. I can look
> at doing that.
>
> I don't like the idea, unless there is any easy to figure way to reformat
> it to default size.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/6449#issuecomment-630370912>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB5HALENUKO2IHCDXRVN2LRSF7I5ANCNFSM4M4ZYJKA>
> .
>


-- 
Thomas Leavitt
Internet enabled since 1990

asvitkine commented:
Actually, testing the behavior even before this PR, the game remembers the window size, even between battles. Just not across launches.

Based on that, I don't think any extra changes are really needed as part of this PR. The behaviour is already good and the small tweak to the starting window size is an improvement over the status quo, but can be refined in later PRs if needed.

DanVanAtta commented:
The fixed location panel for casualties looks good @asvitkine 

Looking at the screenshots for 'total hits', I'm finding I have trouble locating it easily. The recommendation to move to the top was to help in case it was cut-off due to scrolling, what would you think of moving the total hits label into the fixed casualty selection panel?

Second, what do you think of making the 'hit counts' to be red colored to stand out more? IE: making "1 hit" have the '1' be red, or even the whole text of "1 hit" be red.

asvitkine commented:
Moving the hits label into the fixed casualty panel is a bit complex because that panel is a separate component. And we show the number of hits in the history view. So we'd need something like an option on DicePanel to hide its hit label and have the parent component show one instead, which adds a bit of complexity.

Instead, I played around with colors as you suggested and also a different font size. What do you think of this:

![Screen Shot 2020-05-19 at 11 51 55 PM](https://user-images.githubusercontent.com/17648/82403125-1765af80-9a2c-11ea-9f9f-8f6f81a0cddb.png)

DanVanAtta commented:
The larger total hits font looks really good @asvitkine  :+1: 

The red color does make the values stand-out. :+1: 

Did you play around with making the '0 hits' red colored? I think it is very debatable whether that should be colored or not - I'm curious if you gave it a go and had thoughts on it.

I think this is looking quite good @asvitkine. Please create a quick forum post with the above screen-shot if you would, I suspect you'll get very quick feedback. Giving that at least 24 hours to get any late breaking thoughts/concerns would be fair due diligence

@ron-murhammer / @RoiEXLab any additional opinions on the update? 

To summarize,  I think we are quite close @asvitkine  and that the UI is quite solid looking. I'd want to leave it open for just another 24 hours or so to get any final thoughts and feedback before we finalize on this iteration.

RoiEXLab commented on game-core/src/main/java/games/strategy/triplea/ui/DicePanel.java:
```diff
@@ -27,59 +26,82 @@
+  }
+
+  private JLabel makeDiceRolledLabel(final List<Die> dice, final int value) {
+    int hits = 0;
+    for (final Die die : dice) {
+      if (die.getType() == Die.DieType.HIT) {
+        hits++;
+      }
+    }
```
If you want to got for streams
```suggestion
    final long hits = dice.stream().map(Die::getType).filter(Die.DieType.HIT::equals).count();
```

asvitkine replied:
Done. Thanks!

asvitkine commented:
> Did you play around with making the '0 hits' red colored? I think it is very debatable whether that should be colored or not - I'm curious if you gave it a go and had thoughts on it.

I now updated that to make it colored too (it was inconsistent in my previous version because the total 0 was still colored).

Made a post with the changes here:
https://forums.triplea-game.org/topic/1987/improvements-to-combat-ui

asvitkine commented:
One small thing:

Seems the darker red colors look worse on a dark theme (I updated the screenshots in the OP). Do we have a good way to check if the theme is dark or light? If so, I can just use normal red if the theme is dark.

Cernelius commented:
I suggest against any textual colourizations.

asvitkine commented:
> One small thing:
> 
> Seems the darker red colors look worse on a dark theme (I updated the screenshots in the OP). Do we have a good way to check if the theme is dark or light? If so, I can just use normal red if the theme is dark.

Didn't see anything like that already. So I added some logic for it.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/framework/lookandfeel/LookAndFeel.java:
```diff
@@ -78,6 +79,18 @@ public static String getDefaultLookAndFeelClassName() {
+    return background != null && isColorDark(background);
+  }
+
+  public static boolean isColorDark(final Color color) {
```
Would you mind adding some unit testing for this function?

asvitkine replied:
Done.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/ui/DicePanel.java:
```diff
@@ -69,7 +70,9 @@ private JLabel makeDiceRolledLabel(final List<Die> dice, final int value) {
   private static String colorizeHitString(final Object hitsString) {
-    return "<font color='#8B0000'>" + hitsString + "</font>";
+    // On a dark theme, use red. Use a darker red with a light theme.
+    final String color = LookAndFeel.isCurrentLookAndFeelDark() ? "red" : "#8B0000";
```
minor, any objections to extracting "#8B0000" to a named constant? EG: "DARKER_RED"

asvitkine replied:
Done.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/ui/DicePanel.java:
```diff
@@ -27,59 +27,80 @@
+  }
+
+  private void addBottomLabel(final JLabel label) {
+    final GridBagConstraints constraints = new GridBagConstraints();
```
FWIW, there is a type-safe builder for GridBagConstraints that can be used.

Is `gridy` default to `0`, if so, do we need to set `gridx` to `0` as well on line 91?

asvitkine replied:
No, the default value is "RELATIVE", not 0.

asvitkine replied:
Looks like GridBagConstraintsBuilder doesn't support setting gridy to GridBagConstraints.RELATIVE. So let's stick with this. A future PR can always expand the functionality of GridBagConstraintsBuilder to support this use case and refactor this code, if someone wants to contribute (from my perspective, I'm not sure it would add much since the current code isn't very convoluted - and might be more confusing with the extra abstraction).

DanVanAtta replied:
Ah, wasn't aware of the limitation. I'm not sure I'd call the type-safe builders necessarily abstraction; regardless, I do agree the current code is pretty straight forward.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/ui/DicePanel.java:
```diff
@@ -27,59 +27,80 @@
-    add(Box.createVerticalGlue());
-    add(new JLabel("Cost:" + cost));
-    invalidate();
+    addBottomLabel(new JLabel());
```
Any objection to having the bottom label be returned from the function? As is, the method `addBottomLabel` relies on side effects. To understand what is happening, you have to dive into the method, which lacks abstraction. We also have some inconsistency from the other methods that add returned values compared to this last one that adds a component as a side-effect.

asvitkine replied:
Uhm, it's not a side effect, it's the main purpose of the function?

asvitkine replied:
Oh, do you mean the setVerticalAlignment() call?

Given that we no longer use a label with content (before, we used to for total hits) and this is just for taking up the remaining vertical space, we can simplify it. Done.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/ui/history/HistoryDetailsPanel.java:
```diff
@@ -35,32 +38,39 @@
  * </ul>
  */
 public class HistoryDetailsPanel extends JPanel {
-  private static final long serialVersionUID = 5092004144144006960L;
```
nit: changing serialVersionUID strings of swing components should not be necessary. I'm not sure there is consensus whether we should default `serialVersoinId` to a simple `1L`. In essence, I'd recommend to not go out of your way to change them unless it's intentional/necessary.

asvitkine replied:
Reverted. I wonder if there's some project setting we could set to disable warnings about these missing members for swing classes? The only reason we have these useless things is to silence the warnings AFAIK.

asvitkine commented:
Anything else or can this be merged?

DanVanAtta commented:
:+1:
